File: plugins/tokenmonitor/server/py/src/tmon_mcp/state.py
"""Runtime state snapshot exposed by tokenmonitor_status."""

from __future__ import annotations

import threading
import time
from dataclasses import dataclass, field
from enum import Enum

from . import RUNTIME


class Role(str, Enum):
    UNKNOWN = "unknown"
    LEADER = "leader"
    FOLLOWER = "follower"


@dataclass
class UpdateInfo:
    """Cached result of the broker self-version check: is a newer broker/plugin
    release published than the one running. ``known`` is False until the first
    successful fetch of the remote marketplace catalog; while unknown the broker
    advertises nothing (never a false "up to date" or "outdated"). Mirror of Go
    state.UpdateInfo."""

    known: bool = False
    outdated: bool = False
    current: str = ""
    latest: str = ""
    checked_at: str = ""  # RFC3339, empty until the first successful check


@dataclass
class Snapshot:
    runtime: str
    role: str
    role_since: str  # RFC3339
    last_request_at: str = ""
    last_request_remote: str = ""
    last_request_status: int = 0
    requests_total: int = 0
    # update_available is set only once the self-version check has succeeded;
    # None (omitted) means "not yet checked", distinct from False ("up to date").
    update_available: bool | None = None
    latest_version: str = ""

    def to_dict(self) -> dict:
        out = {
            "runtime": self.runtime,
            "role": self.role,
            "role_since": self.role_since,
            "requests_total": self.requests_total,
        }
        if self.last_request_at:
            out["last_request_at"] = self.last_request_at
        if self.last_request_remote:
            out["last_request_remote"] = self.last_request_remote
        if self.last_request_status:
            out["last_request_status"] = self.last_request_status
        if self.update_available is not None:
            out["update_available"] = self.update_available
        if self.latest_version:
            out["latest_version"] = self.latest_version
        return out


def _rfc3339(epoch: float) -> str:
    return time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime(epoch))


@dataclass
class State:
    _role: Role = Role.UNKNOWN
    _role_since: float = field(default_factory=time.time)
    _last_at: float = 0.0
    _last_remote: str = ""
    _last_status: int = 0
    _count: int = 0
    _update: UpdateInfo = field(default_factory=UpdateInfo)
    _lock: threading.Lock = field(default_factory=threading.Lock)

    def set_role(self, role: Role) -> None:
        with self._lock:
            if self._role == role:
                return
            self._role = role
            self._role_since = time.time()

    def record_request(self, remote: str, status: int, when: float | None = None) -> None:
        when = time.time() if when is None else when
        with self._lock:
            self._last_at = when
            self._last_remote = remote
            self._last_status = status
            self._count += 1

    def last_request_at_epoch(self) -> float:
        """When a device last hit the broker (epoch seconds), 0.0 if never.

        The mDNS publisher reads it to decide whether its advertisement has
        gone unheard and should be re-announced.
        """
        with self._lock:
            return self._last_at

    def set_update(self, info: UpdateInfo) -> None:
        """Record the latest broker self-version-check result. The update-check
        poller pokes this concurrently; the broker /sync handler and the MCP
        health/status tools read it back via ``update``."""
        with self._lock:
            self._update = info

    def update(self) -> UpdateInfo:
        """Return the last cached self-version-check result (default =
        known:False, i.e. no check has succeeded yet)."""
        with self._lock:
            return self._update

    def snapshot(self) -> Snapshot:
        with self._lock:
            u = self._update
            return Snapshot(
                runtime=RUNTIME,
                role=self._role.value,
                role_since=_rfc3339(self._role_since),
                last_request_at=_rfc3339(self._last_at) if self._last_at else "",
                last_request_remote=self._last_remote,
                last_request_status=self._last_status,
                requests_total=self._count,
                update_available=(u.outdated if u.known else None),
                latest_version=(u.latest if u.known else ""),
            )

File: plugins/tokenmonitor/server/go/internal/state/state.go
// Package state holds the small piece of runtime state that the MCP tools
// need to introspect: am I the leader or a follower, when did the ESP32
// last hit me, what was the last response code, etc.
//
// Kept deliberately tiny — anything richer (per-IP histograms, latency
// percentiles, …) goes elsewhere. The broker and the leader-election
// loop both poke into the same *State via concurrent-safe setters.
package state

import (
	"sync"
	"time"
)

type Role int

const (
	RoleUnknown Role = iota
	RoleLeader
	RoleFollower
)

func (r Role) String() string {
	switch r {
	case RoleLeader:
		return "leader"
	case RoleFollower:
		return "follower"
	default:
		return "unknown"
	}
}

type State struct {
	mu                sync.RWMutex
	role              Role
	roleSince         time.Time
	lastRequestAt     time.Time
	lastRequestRemote string
	lastRequestStatus int
	requestsTotal     uint64
	update            UpdateInfo
}

// UpdateInfo is the cached result of the broker self-version check: is a newer
// broker/plugin release published than the one running. Known is false until
// the first successful fetch of the remote marketplace catalog; while unknown
// the broker advertises nothing (never a false "up to date" or "outdated").
type UpdateInfo struct {
	Known     bool      `json:"known"`
	Outdated  bool      `json:"outdated"`
	Current   string    `json:"current"`
	Latest    string    `json:"latest,omitempty"`
	CheckedAt time.Time `json:"checked_at,omitempty"`
}

func New() *State {
	return &State{role: RoleUnknown, roleSince: time.Now()}
}

// LastRequestAt reports when a device last hit the broker, zero if never.
// The mDNS publisher reads it to decide whether its advertisement has gone
// unheard and should be re-announced.
func (s *State) LastRequestAt() time.Time {
	s.mu.RLock()
	defer s.mu.RUnlock()
	return s.lastRequestAt
}

// SetRole records a role transition. No-op if the role didn't change.
func (s *State) SetRole(r Role) {
	s.mu.Lock()
	defer s.mu.Unlock()
	if s.role == r {
		return
	}
	s.role = r
	s.roleSince = time.Now()
}

// RecordRequest is called by the broker handler after each /credentials hit
// (regardless of whether auth passed — what matters is observed traffic).
func (s *State) RecordRequest(remote string, status int, t time.Time) {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.lastRequestAt = t
	s.lastRequestRemote = remote
	s.lastRequestStatus = status
	s.requestsTotal++
}

// SetUpdate records the latest broker self-version-check result. The
// update-check poller pokes this concurrently; the broker /sync handler and
// the MCP health/status tools read it back via Update.
func (s *State) SetUpdate(u UpdateInfo) {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.update = u
}

// Update returns the last cached self-version-check result (zero value =
// Known:false, i.e. no check has succeeded yet).
func (s *State) Update() UpdateInfo {
	s.mu.RLock()
	defer s.mu.RUnlock()
	return s.update
}

// Snapshot is an immutable view suitable for serializing back to MCP
// callers. Times are zero values if the event has never happened.
// The Runtime field identifies which language implementation produced
// this snapshot ("go" | "python" | "js"); it lets the launcher and
// users see which port answered without having to introspect $PATH.
type Snapshot struct {
	Runtime           string    `json:"runtime"`
	Role              string    `json:"role"`
	RoleSince         time.Time `json:"role_since"`
	LastRequestAt     time.Time `json:"last_request_at,omitempty"`
	LastRequestRemote string    `json:"last_request_remote,omitempty"`
	LastRequestStatus int       `json:"last_request_status,omitempty"`
	RequestsTotal     uint64    `json:"requests_total"`
	// UpdateAvailable is true when the self-version check found a newer
	// broker/plugin release than the one running. Omitted (absent) until the
	// first successful check, so callers can distinguish "up to date" from
	// "not yet checked".
	UpdateAvailable *bool  `json:"update_available,omitempty"`
	LatestVersion   string `json:"latest_version,omitempty"`
}

// Runtime is the identifier this implementation reports in the
// Snapshot. The Go impl is "go"; Python/JS ports must report their own
// values to keep tokenmonitor_status diagnosable end-to-end.
const Runtime = "go"

func (s *State) Snapshot() Snapshot {
	s.mu.RLock()
	defer s.mu.RUnlock()
	snap := Snapshot{
		Runtime:           Runtime,
		Role:              s.role.String(),
		RoleSince:         s.roleSince,
		LastRequestAt:     s.lastRequestAt,
		LastRequestRemote: s.lastRequestRemote,
		LastRequestStatus: s.lastRequestStatus,
		RequestsTotal:     s.requestsTotal,
	}
	if s.update.Known {
		avail := s.update.Outdated
		snap.UpdateAvailable = &avail
		snap.LatestVersion = s.update.Latest
	}
	return snap
}
